urwid_readline/readline_edit.py

```diff
@@ -338,17 +342,23 @@ def insert_new_line(self):
         if self.multiline:
             self.insert_text("\n")
 
-    def enable_autocomplete(self, func, key=None):
+    def enable_autocomplete(self, func, key=None, key_reverse=None):
         self._autocomplete_func = func
         self._autocomplete_key = key if key is not None else "tab"
+        self._autocomplete_key_reverse = key_reverse if key_reverse is not None else "shift tab"
         self._autocomplete_state = None
 
     def set_completer_delims(self, delimiters):
         self._autocomplete_delims = delimiters
 
-    def _complete(self):
+    def _complete(self, cycle_forward):
         if self._autocomplete_state:
-            self._autocomplete_state.num += 1
+            if self._autocomplete_state.num == 0 and not cycle_forward:
+                self._autocomplete_state.num = None
+            elif self._autocomplete_state.num == -1 and cycle_forward:
+                self._autocomplete_state.num = None
```

Comment:
Taking care of a couple of cases manually:
When you are at state 0 and cycle backwards, you would ideally want to get the original/non-autocompleted text. Setting state to -1 would set it to the last suggestion.
Same for when you're at -1 and cycle forward

Reply:
Returning None causes a `TypeError` while indexing, so I've taken care of it there. I'm not entirely sure if this is the best way to handle the exception, so let me know if it's okay.